angr/state_plugins/utils.py

```diff
@@ -0,0 +1,90 @@
+import angr
+import sys
+
+import claripy
+
+
+def get_unconstrained_bytes(state, name, bits, source=None, memory=None):
+
+    if (memory is not None and memory.category == 'mem' and
+                angr.options.CGC_ZERO_FILL_UNCONSTRAINED_MEMORY in state.options):
```

Comment:
I'm guessing this code was written before we added the complex warnings about the sources of uninitialized memory access? This definitely does not mesh with the current logic for zero-fill options

Reply:
Yep, this was written a long time ago, before the switch to python3.

This shouldn't be hard to fix. I think we just need to mimic the warning message, and check for options.SYMBOL_FILL_UNCONSTRAINED_MEMORY vs ZERO_FILL_UNCONSTRAINED_MEMORY here instead of/in addition to angr.options.CGC_ZERO_FILL_UNCONSTRAINED_MEMORY, right?

Reply:
Yes, but you also need to remove all other similar logic in angr and replace them with this helper function.

I'm worried that the logic has changed drastically enough that the abstraction you've broken out won't be sufficient, though.

Reply:
The current logic should be workable. I'm just curious about what UNDER_CONSTRAINED_SYMEXEC really does (used in SimSymbolicMemory.get_unconstrained_bytes()).

Specifically, I don't want to needlessly tick uc_alloc_depth on every load (which is what would happen now).

In the paper, before introducing an unconstrained symbolic variable the algorithm checks if it is _sat_ that a load would ever yield an uninit read. I'm not sure why the reference implementation doesn't do this--perf? Either way, if I add in this check, we won't tick uc_alloc_depth on every load and I can more easily mimic the uninit log statement in SimSymbolicMemory._fill_missing()

Reply:
Done. I added in the SAT check to prevent log spam, though I noticed a small perf hit (of course). There's also a typo in the original paper I think. I'm ignoring UNDER_CONSTRAINED_SYMEXEC for now.